Upgrade our TFM to net9.0

jkoritzinsky opened the pull request:
Upgrade the TFM in the repo to net9.0

Drop support for net6.0 and net7.0

Follow-up to #90558

ViktorHofer commented on src/installer/tests/Assets/TestUtils/TestProjects.targets:
```diff
@@ -6,25 +6,25 @@
     <KnownFrameworkReference Include="Microsoft.NETCore.App"
-                             DefaultRuntimeFrameworkVersion="7.0.0"
-                             LatestRuntimeFrameworkVersion="7.0.0"
+                             DefaultRuntimeFrameworkVersion="9.0.0"
+                             LatestRuntimeFrameworkVersion="9.0.0"
```
These versions refer to the latest available targeting and runtime pack, i.e. 8.0.0-rc1....

ViktorHofer commented on eng/targetingpacks.targets:
```diff
@@ -115,6 +115,8 @@
                        Condition="'$(UsePackageDownload)' == 'true' and $([System.String]::Copy('%(Identity)').EndsWith('Microsoft.DotNet.ILCompiler'))" />
       <PackageReference Remove="@(PackageReference)"
                         Condition="'$(UsePackageDownload)' != 'true' and $([System.String]::Copy('%(Identity)').EndsWith('Microsoft.DotNet.ILCompiler'))" />
+      <PackageReference Remove="@(PackageReference)"
+                        Condition="'%(Identity)' == 'Microsoft.DotNet.ILCompiler'" />
```
Shouldn't this already be covered by the above four lines? If not, should we change the above four lines instead?

jkoritzinsky replied:
The hosting pack (`runtime.win-x64.Microsoft.DotNet.ILCompiler`) and the like are covered by the above cases. However, `Microsoft.DotNet.ILCompiler` is always referenced as a PackageReference it seems. As a result, when `UsePackageDownload` is true, we need this case as well.

ViktorHofer commented on src/mono/sample/iOS-NativeAOT/Program.csproj:
```diff
@@ -45,9 +45,9 @@
+      <KnownRuntimePack Condition="%(RuntimePackLabels) == 'NativeAOT'" LatestRuntimeFrameworkVersion="9.0.0-dev" />
       <KnownILCompilerPack Update="Microsoft.DotNet.ILCompiler">
-        <ILCompilerPackVersion>8.0.0-dev</ILCompilerPackVersion>
+        <ILCompilerPackVersion>9.0.0-dev</ILCompilerPackVersion>
```
```suggestion
        <ILCompilerPackVersion>$(MajorVersion).0.0-dev</ILCompilerPackVersion>
```

ViktorHofer commented on src/libraries/Microsoft.Bcl.Cryptography/src/Microsoft.Bcl.Cryptography.csproj:
```diff
@@ -1,13 +1,9 @@
 
   <PropertyGroup>
-    <TargetFrameworks>$(NetCoreAppCurrent);netstandard2.0;$(NetFrameworkMinimum)</TargetFrameworks>
+    <TargetFrameworks>$(NetCoreAppCurrent);netstandard2.0;$(NetCoreAppPrevious);$(NetCoreAppMinimum);$(NetFrameworkMinimum)</TargetFrameworks>
```
We care about the TFM order here as that's how VS displays the inner builds:

```suggestion
    <TargetFrameworks>$(NetCoreAppCurrent);$(NetCoreAppPrevious);$(NetCoreAppMinimum);netstandard2.0;$(NetFrameworkMinimum)</TargetFrameworks>
```

ViktorHofer commented on src/libraries/Microsoft.Bcl.TimeProvider/src/Microsoft.Bcl.TimeProvider.csproj:
```diff
@@ -1,12 +1,8 @@
 
   <PropertyGroup>
-    <TargetFrameworks>$(NetCoreAppCurrent);netstandard2.0;$(NetFrameworkMinimum)</TargetFrameworks>
+    <TargetFrameworks>$(NetCoreAppCurrent);netstandard2.0;$(NetCoreAppPrevious);$(NetCoreAppMinimum);$(NetFrameworkMinimum)</TargetFrameworks>
```
same comment about TFM order:

```suggestion
    <TargetFrameworks>$(NetCoreAppCurrent);$(NetCoreAppPrevious);$(NetCoreAppMinimum);netstandard2.0;$(NetFrameworkMinimum)</TargetFrameworks>
```

ViktorHofer commented on src/mono/msbuild/apple/build/AppleBuild.LocalBuild.props:
```diff
@@ -22,9 +22,9 @@
+    <_NetCoreAppCurrent>net9.0</_NetCoreAppCurrent>
     <TargetFramework>$(_NetCoreAppCurrent)</TargetFramework>
-    <_NetCoreAppToolCurrent>$(_NetCoreAppCurrent)</_NetCoreAppToolCurrent>
+    <_NetCoreAppToolCurrent>net8.0</_NetCoreAppToolCurrent>
```
Please add TODO comment here to indicate that this should be updated when a .NET 9 SDK is consumed that supports the net9.0 TFM.

fanyang-mono committed:
Update TFM to net9.0

jkoritzinsky committed:
Fix ApiCompat failures.

jkoritzinsky committed:
Create a separate workload for net9 vs net8. Let's see if this fixes WASI.

jkoritzinsky commented:
@radical can you take a look at the workload changes I made? I'm trying to get the WBT test suites passing. These were/are the last pieces failing in this PR.

radical commented:
> @radical can you take a look at the workload changes I made? I'm trying to get the WBT test suites passing. These were/are the last pieces failing in this PR.

Sure, I can take a look later today.

ViktorHofer commented on src/mono/wasm/testassets/WasmBasicTestApp/WasmBasicTestApp.csproj:
```diff
@@ -1,6 +1,6 @@
 <Project Sdk="Microsoft.NET.Sdk.WebAssembly">
   <PropertyGroup>
-    <TargetFramework>net8.0</TargetFramework>
+    <TargetFramework>net9.0</TargetFramework>
```
Is there a reason why this project can't use `$(NetCoreAppCurrent)`?

radical replied:
We can do that.

ViktorHofer commented on src/coreclr/tools/aot/ILCompiler.Reflection.ReadyToRun/ILCompiler.Reflection.ReadyToRun.csproj:
```diff
@@ -16,6 +16,11 @@
     <Platforms>AnyCPU;x64</Platforms>
     <PlatformTarget>AnyCPU</PlatformTarget>
     <RunAnalyzers>false</RunAnalyzers>
+    <!--
+      The SDK warns if a project targets a framework that doesn't support trimming and IsTrimmable is true.
+      We set IsTrimmable higher up in the Directory.Build.props files, so we reset it to false here.
+    -->
+    <IsTrimmable>false</IsTrimmable>
```
Are you sure this is still necessary? From what I can see, the IsTrimmable=true property is now only set for .NETCoreApp under tools/aot.

radical commented:
> > @radical can you take a look at the workload changes I made? I'm trying to get the WBT test suites passing. These were/are the last pieces failing in this PR.
> 
> Sure, I can take a look later today.

I'll have to dig in tomorrow.

radical committed:
revert all wasm changes

radical commented:
re:wasm, we can't change the tfm for wasm.build.tests yet, because these tests *build* projects with the sdk which needs to start supporting `tfm=net9.0`. Additionally emsdk needs new packages, and then corresponding we would add packages in runtime. For now, we'll need some little hack to get this passing, but it won't strictly be testing "the right stuff".

I'm working on this in https://github.com/dotnet/runtime/pull/91105 .

ViktorHofer committed:
Remove netcoreappX.X, net6.0, net7.0 and net8.0 conditions and clean-up

ViktorHofer committed:
Merge branch 'main' into net90tfm

ViktorHofer commented:
@jkoritzinsky I pushed to your branch to update the project files under src/libraries to remove any non-necessary netcoreappX.X, net6.0, net7.0 or net8.0 conditions. We have been doing that always in conjunction with updating the TFM. Hope that's OK.

ViktorHofer committed:
Fix build

ViktorHofer committed:
Make projects target the right set of TFMs and order

ViktorHofer commented on src/libraries/Microsoft.Extensions.Caching.Abstractions/src/CompatibilitySuppressions.xml:
```diff
@@ -0,0 +1,32 @@
+    <Right>right</Right>
+    <IsBaselineSuppression>true</IsBaselineSuppression>
+  </Suppression>
+  <Suppression>
+    <DiagnosticId>CP1002</DiagnosticId>
+    <Target>System.Runtime, Version=4.1.2.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a</Target>
+    <Left>lib/net7.0/Microsoft.Extensions.Caching.Abstractions.dll</Left>
+    <Right>right</Right>
+    <IsBaselineSuppression>true</IsBaselineSuppression>
+  </Suppression>
```
We need to understand these suppressions. I'm not sure which assembly reference this refers to.

ViktorHofer commented on global.json:
```diff
@@ -1,11 +1,11 @@
   },
   "tools": {
-    "dotnet": "8.0.100-preview.7.23376.3"
+    "dotnet": "8.0.100-rc.1.23415.11"
```
see response in https://github.com/dotnet/runtime/issues/90434#issuecomment-1693423572

ViktorHofer committed:
Update assembly version

ViktorHofer committed:
Remove dead property

radical committed:
[wasm] WBT: Override runtime pack with the locally built one

radical committed:
Merge branch 'main' into net90tfm

ViktorHofer commented:
wow, looks like there are only a 4 more failing legs, all wasm.

radical committed:
[wasm] Move the override targets to a separate file, and use for wasi too

radical committed:
overwrite when copying

radical committed:
add missing file

ViktorHofer commented on src/mono/wasm/Wasm.Build.Tests/data/WasmOverridePacks.targets:
```diff
@@ -0,0 +1,63 @@
+<Project>
```
Everything in this file is a copy of targetingpacks.targets. Why is that?

radical replied:
This was added in the wasm.build.tests quite some time ago, and then removed. I can try using `targetingpacks.targets`. I can make a copy of it at build time, to use with test project projects that are generated, and built for individual tests.

ViktorHofer replied:
Not required for this PR but it would be good to file a follow-up issue to remove the duplication.

radical replied:
https://github.com/dotnet/runtime/issues/91350

radical committed:
Update WebAssembly.Pack version also

radical committed:
fix tests

radical committed:
Disable non-wasm build tests

Trying to get this to work for the short period till the sdk moves to
net9.0 would not be worth the effort. Failure:

```
        [] Executing (Captured Output) - /root/helix/work/workitem/e/dotnet-latest/dotnet run -c Debug -f net8.0 --no-build -  in pwd /root/helix/work/workitem/e/wbt/nonwasm_net8.0_Debug_4utlyi5x_k3s
        [] Unhandled exception. System.IO.FileNotFoundException: Could not load file or assembly 'System.Runtime, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.
```

radical committed:
Override runtime pack for multithread

radical committed:
[wasm] more fixes for wbt

radical committed:
[wasm] override packs only when using workloads

jkoritzinsky commented:
I think we've got everything fixed up here excluding known issues. @ViktorHofer any more feedback?

ViktorHofer commented:
Yeah, I need to debug one of those new suppressions. They shouldn't show up like that.

marek-safar commented on src/installer/tests/Assets/TestUtils/TestProjects.props:
```diff
@@ -5,7 +5,7 @@
 
   <PropertyGroup>
-    <NetCoreAppCurrent>net8.0</NetCoreAppCurrent>
+    <NetCoreAppCurrent>net9.0</NetCoreAppCurrent>
```
Why not to use NetCoreAppCurrentVersion?

ViktorHofer replied:
These projects don't import the repo infrastructure.

ViktorHofer commented:
Dropping TFMs is a breaking change. We need to file a breaking change doc before we release .NET 9.
